### PR TITLE
#3076 Swapped sn/lot details at bottom in log section

### PR DESCRIPTION
**Pull request by jrbirch99**
- Add messages to thread of picking showing tracked products were swapped
- Update unit tests for different edge cases

### CODE REVIEW

File: addons/udes_stock/models/stock_move_line.py [hunk 1]
```diff
@@ -504,6 +522,23 @@ def _swap_tracked_items(
...
                 # Remove lot names that have been processed so they are not called below
                 lot_names = list(set(lot_names) - set(lot_names_swapped_in))
+
+            # For the message thread to notify of the tracked products swapped in
+            new_mls = StockMoveLine.browse() 
+            for ml, mls_values in res.items():
+                new_mls = new_mls | ml
+            
+            for picking, mls in new_mls.groupby("picking_id"):
+                swap_in_title = "Tracked Products Swapped In: <ul>"
```
Comment (kdwyer) [inline addons/udes_stock/models/stock_move_line.py:552]: Likewise, this can be outside the loop?

File: addons/udes_stock/models/stock_move_line.py [hunk 2]
```diff
@@ -425,6 +425,24 @@ def _swap_tracked_items(
...
 
+                # For the message thread to notify of the tracked product swapped out
+                old_mls = mls_need_changing | swapped_lot_on_assigned_mls
+                for picking, mls in old_mls.groupby("picking_id"): 
+                    picking_qty = quantity_swapping
+                    swap_out_title = "Tracked Products Swapped Out: <ul>"
+                    swap_out_text=""
+                    for ml in mls:
+                        picking_qty -= ml.product_uom_qty
+                        swapped_qty = min([quantity_swapping,ml.product_uom_qty])
```
Comment (PeterAlabaster) [inline addons/udes_stock/models/stock_move_line.py]: nit, space between `,` - mistook it for a dot

**Review by kdwyer — APPROVED** [review]

File: addons/udes_stock/models/stock_move_line.py [hunk 3]
```diff
@@ -425,6 +425,24 @@ def _swap_tracked_items(
                     ], order="id"
                 )
 
+                # For the message thread to notify of the tracked product swapped out
+                old_mls = mls_need_changing | swapped_lot_on_assigned_mls
+                for picking, mls in old_mls.groupby("picking_id"): 
+                    picking_qty = quantity_swapping
+                    swap_out_title = "Tracked Products Swapped Out: <ul>"
```
[the remark this answers is not in the document]
Comment (PeterAlabaster) [reply addons/udes_stock/models/stock_move_line.py:435]: the `end_of_list` variable too.

It might be worth just defining them inline at the end and not bother making them variable (as they aren't ever changed):
```python
body = "Tracked Products Swapped Out: <ul>" + swap_out_text + "</ul>"
```

File: addons/udes_stock/models/stock_move_line.py (hunk 1 above)
Comment (PeterAlabaster) [inline addons/udes_stock/models/stock_move_line.py:552]: Same nit as kevin here

Comment (PeterAlabaster) [reply addons/udes_stock/models/stock_move_line.py:552]: kevin already raised. resolving as duplicate

File: addons/udes_stock/models/stock_move_line.py (hunk 2 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py]: this was added in during the formatting commit :slightly_smiling_face:

File: addons/udes_stock/models/stock_move_line.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py]: Changed, used the inline suggestion :slightly_smiling_face:

File: addons/udes_stock/models/stock_move_line.py (hunk 1 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py]: changed

**Review by PeterAlabaster — APPROVED** [review]

File: addons/udes_stock/models/stock_move_line.py (hunk 2 above)
Comment (PeterAlabaster) [reply addons/udes_stock/models/stock_move_line.py]: Ah my bad, was looking at the commit on its own cause black formatting is always hard to see what's changed :+1:

File: addons/udes_stock/models/stock_move_line.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PeterAlabaster) [reply addons/udes_stock/models/stock_move_line.py]: I'm happy, will let @kdwyer resolve if happy with inline
